*deprecated* [#480] Grouping Consecutive Ping Notifications

tyzhou05 opened the pull request:
## Issue Number:

## Self Check List

- [x] Have you double-checked the base branch into which this PR will be merged?
- [x] Have you rebased from the base branch?

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (formatting, renaming)
- [ ] Merge Develop to Main
- [ ] Other (please describe):

## What does this PR do?
#480

### Implementation
- Since the clutter of notifications occurs in a specific timeframe, I created a `find_recent_ping` class to handle this.
- Implementation groups all notifications created within 5 minutes into a group and updates the message with the number of pings sent in that time frame. I tested with 10 seconds, but I think 5 minutes or longer makes sense for the actual deployment -- @njs03332 let me know thoughts!

### Testing
- First, login to adoor_2 via user/login:
  - [POST api/user/login/](https://diivers.postman.co/workspace/WhoAmI~c2317c3a-2ab1-458e-96b5-01f8ca6b983e/request/6673035-c2378508-e641-475d-a68e-09b3ffeb6dc5?action=share&source=copy-link&creator=39127770)
- Next, login to adoor_3, adoor_1, and adoor_4 via user/login and then POST pings repeatedly to adoor_2's user id (3).
  - [POST api/ping/user/int:pk/](https://diivers.postman.co/workspace/WhoAmI~c2317c3a-2ab1-458e-96b5-01f8ca6b983e/request/6673035-02539b66-fb2e-4235-9d22-bb88369c1226?action=share&source=copy-link&creator=39127770)
- Wait 5 minutes and then log back into adoor_2 and view the pings grouped together under one notification:
  - [GET api/notifications](https://diivers.postman.co/workspace/WhoAmI~c2317c3a-2ab1-458e-96b5-01f8ca6b983e/request/6673035-ea57105b-5076-4653-a94d-0d96204f262f?action=share&source=copy-link&creator=39127770)
```
  {
            "id": 206,
            "is_response_request": false,
            "is_friend_request": false,
            "recent_actors": [
                {
                    "id": 2,
                    "username": "adoor_1",
                    "profile_pic": "#66A6A7",
                    "url": "http://localhost:8000/api/user/adoor_1/profile/",
                    "profile_image": null
                },
                {
                    "id": 2,
                    "username": "adoor_1",
                    "profile_pic": "#66A6A7",
                    "url": "http://localhost:8000/api/user/adoor_1/profile/",
                    "profile_image": null
                },
                {
                    "id": 2,
                    "username": "adoor_1",
                    "profile_pic": "#66A6A7",
                    "url": "http://localhost:8000/api/user/adoor_1/profile/",
                    "profile_image": null
                }
            ],
            "notification_type": "Ping",
            "is_recent": true,
            "message": "adoor_1 sent you 3 pings",
            "question_content": null,
            "is_read": true,
            "created_at": "2025-02-25T10:11:27.873334-08:00",
            "redirect_url": "/users/adoor_1/ping",
            "notification_updated_at": "2025-02-25T10:11:28.990282-08:00"
        },

```
Tested several additional common / edge cases (e.g. send group pings from multiple users within 5 minutes; send separate group pings from different users and verified they are created correctly within 5 minutes; mark all notifications as read and then send group pings -> they were not grouped with read notifications)

## Preview Image

## Further comments

tyzhou05 committed:
define get_recent_ping; added cutoff time of 24hours

tyzhou05 committed:
editing implementation + testing

tyzhou05 committed:
fix ping_count not use actors

fixing small bug where notif would say # of actors rather than # of pings

tyzhou05 committed:
fix ping_count not use actors

fixing small bug where notif would say # of actors rather than # of pings

njs03332 commented on adoorback/notification/models.py:
```diff
@@ -31,57 +31,61 @@ def admin_only(self, **kwargs):
                                       emoji=None):
         noti_to_update = None
 
-        if target.type == "Like":
```
Hmm why did you delete the original code related to like, response request and reaction?

tyzhou05 replied:
Ah I think I must have accidentally done so while implementing - so sorry! I just reverted it and added the changes, and I believe the functionality won't change the Postman testing cases but I will test again as well

tyzhou05 committed:
refactor: #480 reinstate like, response request and reaction

tyzhou05 committed:
refactor: #480 reinstate like, response request and reaction

njs03332 commented:
Hmm I'm getting this error when I send my second ping!
```shell
2025-03-02 09:06:26 ERROR HTTP POST /api/ping/user/2/ 500 [0.12, 127.0.0.1:49652]
2025-03-02 09:07:22 ERROR Internal Server Error: /api/ping/user/2/
Traceback (most recent call last):
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/asgiref/sync.py", line 518, in thread_handler
    raise exc_info[1]
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/django/core/handlers/exception.py", line 42, in inner
    response = await get_response(request)
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/django/core/handlers/base.py", line 253, in _get_response_async
    response = await wrapped_callback(
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/asgiref/sync.py", line 468, in __call__
    ret = await asyncio.shield(exec_coro)
  File "/Users/yuri/.pyenv/versions/3.9.7/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/asgiref/sync.py", line 522, in thread_handler
    return func(*args, **kwargs)
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/django/views/decorators/csrf.py", line 56, in wrapper_view
    return view_func(*args, **kwargs)
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/django/views/generic/base.py", line 104, in view
    return self.dispatch(request, *args, **kwargs)
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/rest_framework/views.py", line 506, in dispatch
    response = handler(request, *args, **kwargs)
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/rest_framework/generics.py", line 246, in post
    return self.create(request, *args, **kwargs)
  File "/Users/yuri/Documents/git/WhoAmI-Today-backend/adoorback/ping/views.py", line 67, in create
    response = super().create(request, *args, **kwargs)
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/rest_framework/mixins.py", line 19, in create
    self.perform_create(serializer)
  File "/Users/yuri/Documents/git/WhoAmI-Today-backend/adoorback/ping/views.py", line 64, in perform_create
    serializer.save(sender=user, receiver=connected_user, ping_room=ping_room)
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/rest_framework/serializers.py", line 208, in save
    self.instance = self.create(validated_data)
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/rest_framework/serializers.py", line 989, in create
    instance = ModelClass._default_manager.create(**validated_data)
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/django/db/models/query.py", line 658, in create
    obj.save(force_insert=True, using=self.db)
  File "/Users/yuri/Documents/git/WhoAmI-Today-backend/adoorback/ping/models.py", line 110, in save
    super().save(*args, **kwargs)
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/safedelete/models.py", line 131, in save
    super(SafeDeleteModel, self).save(**kwargs)
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/django/db/models/base.py", line 814, in save
    self.save_base(
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/django/db/models/base.py", line 892, in save_base
    post_save.send(
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/django/dispatch/dispatcher.py", line 176, in send
    return [
  File "/Users/yuri/.pyenv/versions/whoami/lib/python3.9/site-packages/django/dispatch/dispatcher.py", line 177, in <listcomp>
    (receiver, receiver(signal=self, sender=sender, **named))
  File "/Users/yuri/Documents/git/WhoAmI-Today-backend/adoorback/ping/models.py", line 142, in create_ping_notification
    Notification.objects.create_or_update_notification(
  File "/Users/yuri/Documents/git/WhoAmI-Today-backend/adoorback/notification/models.py", line 73, in create_or_update_notification
    ping_count = noti_to_update.target_set.count() + 1
AttributeError: 'Notification' object has no attribute 'target_set'
2025-03-02 09:07:22 ERROR HTTP POST /api/ping/user/2/ 500 [0.14, 127.0.0.1:49670]
```

tyzhou05 committed:
refactor: #480 fix internal server error

fix: #480 rebase errors#

tyzhou05 committed:
refactor: #480 fix internal server error

tyzhou05 commented:
I've added more functionality to my implementation and tested the cases and it works again as intended! I think it was some small changes necessary with calling the correct actors and handling ValidationErrors to find the most recent notification.

Let me know if it doesn't look good from your side! @njs03332

tyzhou05 committed:
remove unnecessary migration file

remove: #480 irrelevant migration file#

tyzhou05 committed:
remove unnecessary migration file

tyzhou05 committed:
fix: #480 merge conflicts

tyzhou05 committed:
fix: #480 rebase issues

tyzhou05 commented:
@njs03332 I was just fixing some rebase issues I noticed, but it seems that I inadvertently might have broken the implementation for the grouping ping notifications - will work on this and see what's going on!

njs03332 committed:
(#520) add load_questions command (#539)

feat: #520 add load_questions command

njs03332 committed:
(#535) change ping notification redirect url (#540)

bugfix: #535 change ping notification redirect url

njs03332 committed:
(#534) set notification message limit to 300 (#541)

* bugfix: #534 set notification message limit to 300

* bugfix: #534 add migration file

njs03332 committed:
(#530) fix redirect url for 'became friends' notification when username changed (#543)

bugfix: #530 fix redirect url for 'became friends' notification when username changed

njs03332 committed:
(#527) exclude superuser and friend request received users from recommended friends (#544)

refactor: #527 exclude superuser and friend request received users from recommended friends

njs03332 committed:
(#529) fix post_delete FriendRequest method to post_save (#545)

* bugfix: #529 fix post_delete FriendRequest method to post_save (due to safe delete)

* bugfix: #529 remove unnecessary log printing

njs03332 committed:
(#531) make requestee_choice optional in ver.Q friend request update api (#549)

bugfix: #531 make requestee_choice optional in ver.Q friend request update api

ssm0318 committed:
(#526) update strings (#550)

* update strings

* fix minor error

njs03332 committed:
(#553) filter feed post list using is_audience (#554)

bugfix: #553 filter feed post list using is_audience

njs03332 committed:
(#552) raise error instead of returning PermissionDenied object (#555)

bugfix: #552 raise error instead of returning PermissionDenied object in get_queryset()

njs03332 committed:
(#547) mark all notes in feed as read (#556)

feat: #547 mark all notes in feed as read

njs03332 commented:
Has this branch been rebased yet? 👀

tyzhou05 committed:
[#496] Functionality for update_past_posts in ver.Q (#532)

* feat: #496 add functionality for update_past_posts

* (#506) modify get_object() return value of end & touch session APIs (#507)

bugfix: #506 modify get_object() return value of end & touch session APIs

* refactor: #496 splitting update_past_posts into requester and requestee; adding Connection model implementation and updates to save() function

* refactor: #496 serializers not displaying requestee and requester update_past_posts fields properly

* refactor: #496 removing redundant methods, refactoring implementation

* refactor: #496 refactor requestee/requester logic in serializers

* fix: #496 removing unnecessary serializer declarations

* fix: #496 removing last redundant instance of requester_update_posts

---------

Co-authored-by: Yuri <37107740+njs03332@users.noreply.github.com>

tyzhou05 committed:
Merge branch 'main' into 480

tyzhou05 committed:
fix: #480 rebase errors; fixing actor errors with ping group notification, resolve message_ko/message_en bug

tyzhou05 commented:
@njs03332 So sorry, I thought I had fixed the rebase errors earlier but I still think there were some more! I rebased again, made some changes to the grouping by actors functionality, and after testing I believe the original Ping notification grouping I created is back to working as intended. 

However, I think the message while the notification updates will always just show as `adoor_3 sent you a message` instead of `adoor_3 sent you x pings`. Not sure if this phrasing would be okay - I think I can change the implementation to be more specific about pings, but let me know if the grouping functionality looks good from your end!

```
{
    "count": 38,
    "next": "http://localhost:8000/api/notifications/?page=2",
    "previous": null,
    "results": [
        {
            "id": 197,
            "is_response_request": false,
            "is_friend_request": false,
            "recent_actors": [
                {
                    "id": 4,
                    "username": "adoor_3",
                    "profile_pic": "#9188CC",
                    "url": "http://localhost:8000/api/user/adoor_3/profile/",
                    "profile_image": null
                },
                {
                    "id": 4,
                    "username": "adoor_3",
                    "profile_pic": "#9188CC",
                    "url": "http://localhost:8000/api/user/adoor_3/profile/",
                    "profile_image": null
                },
                {
                    "id": 4,
                    "username": "adoor_3",
                    "profile_pic": "#9188CC",
                    "url": "http://localhost:8000/api/user/adoor_3/profile/",
                    "profile_image": null
                },
                {
                    "id": 4,
                    "username": "adoor_3",
                    "profile_pic": "#9188CC",
                    "url": "http://localhost:8000/api/user/adoor_3/profile/",
                    "profile_image": null
                },
                {
                    "id": 4,
                    "username": "adoor_3",
                    "profile_pic": "#9188CC",
                    "url": "http://localhost:8000/api/user/adoor_3/profile/",
                    "profile_image": null
                },
                {
                    "id": 4,
                    "username": "adoor_3",
                    "profile_pic": "#9188CC",
                    "url": "http://localhost:8000/api/user/adoor_3/profile/",
                    "profile_image": null
                }
            ],
            "notification_type": "Ping",
            "is_recent": true,
            "message": "adoor_3 sent you a message!",
            "question_content": null,
            "is_read": false,
            "created_at": "2025-03-13T10:55:09.381200-07:00",
            "redirect_url": "/users/adoor_3/ping",
            "notification_updated_at": "2025-03-13T10:56:34.541999-07:00"
        },
```

njs03332 commented:
Nice! I thought I should get some other opinions on this, so I asked around. I'll let you know as soon as I get a response! :)

njs03332 commented on adoorback/ping/migrations/0002_alter_ping_content.py:
```diff
@@ -1,19 +0,0 @@
-# Generated by Django 4.2.14 on 2025-03-01 05:31
```
This migration file is from another PR! You shouldn't delete this one

njs03332 commented on adoorback/notification/models.py:
```diff
@@ -176,9 +223,6 @@ class NotificationActor(AdoorTimestampedModel, SafeDeleteModel):
     class Meta:
         ordering = ['-created_at']
 
-    def __str__(self):
```
Why did you delete this part?

njs03332 commented on adoorback/notification/models.py:
```diff
@@ -82,10 +96,29 @@ def create_or_update_notification(self, actor, user, origin, target, noti_type,
+                updated_message_en = f"{actor.username} sent you {ping_count} pings"
+                noti_to_update.message_ko = updated_message_ko
+                noti_to_update.message_en = updated_message_en
+                noti_to_update.message = updated_message_en
```
This line is unnecessary. If you set `message_ko` and `message_en`, Django will automatically assign the value to `message` based on the default language. (With our current specifications, it is being set to Korean.)

njs03332 commented on adoorback/notification/models.py:
```diff
@@ -94,24 +127,38 @@ def create_or_update_notification(self, actor, user, origin, target, noti_type,
+                redirect_url=redirect_url,
+                message_ko=message_ko,
+                message_en=message_en,
+                message=message_en
```
This line should be deleted too.

njs03332 commented on adoorback/notification/models.py:
```diff
@@ -82,10 +96,29 @@ def create_or_update_notification(self, actor, user, origin, target, noti_type,
+                NotificationActor.objects.create(user=actor, notification=noti_to_update)
+                ping_count = noti_to_update.actors.count()
+                
+                updated_message_ko = f"{actor.username}님이 {ping_count}번 핑을 보냈습니다"
+                updated_message_en = f"{actor.username} sent you {ping_count} pings"
```
```suggestion
                updated_message_ko = f"{actor.username}님이 {ping_count}번 메시지를 보냈습니다!"
                updated_message_en = f"{actor.username} sent you {ping_count} messages!"
```

njs03332 commented on adoorback/notification/models.py:
```diff
@@ -94,24 +127,38 @@ def create_or_update_notification(self, actor, user, origin, target, noti_type,
         else:
-            message_ko, message_en = construct_message(noti_type, actor.username + "님", None,
+            if target._meta.model_name == 'ping':
+                message_ko = f"{actor.username}님이 핑을 보냈습니다"
+                message_en = f"{actor.username} sent you a Ping!"
```
We renamed ping to message! 
This should be changed to 
```suggestion
                message_ko = f"{actor.username}님이 메시지를 보냈습니다!"
                message_en = f"{actor.username} sent you a message!"
```

njs03332 commented on adoorback/ping/models.py:
```diff
@@ -139,12 +139,19 @@ def create_ping_notification(created, instance, **kwargs):
     )
-    NotificationActor.objects.create(user=sender, notification=noti)
+
+    if noti:
+        noti.message_ko = f"{sender.username}님이 메시지를 보냈습니다!"
+        noti.message_en = f"{sender.username} sent you a message!"
+        noti.message = noti.message_en 
+        noti.save()
```
I think you need to get rid of this part! It's overriding the message of notification that you originally modified
(`{actor.username} sent you {ping_count} pings`)

njs03332 commented on adoorback/notification/serializers.py:
```diff
@@ -30,8 +30,14 @@ def get_is_friend_request(self, obj):
         from account.serializers import UserMinimalSerializer
-        recent_actors = obj.actors.all()[:3]
-        return UserMinimalSerializer(recent_actors, many=True).data
+        if obj.target and obj.target._meta.model_name == 'ping':
+            actor = obj.actors.first()
+            count = obj.actors.count()
+            actors = [actor] * count 
+            return UserMinimalSerializer(actors, many=True).data
```
`recent_actors` field is used on the frontend side to display profile images in the notification.
It's a bit awkward to display 3 same images, so I think it's better to return only one actor.
<img width="493" alt="스크린샷 2025-03-16 오후 2 54 20" src="https://github.com/user-attachments/assets/85358a00-2ff1-4e4a-adda-52a6986db953" />

```suggestion
        if obj.target and obj.target._meta.model_name == 'ping':
            return UserMinimalSerializer([obj.actors.first()], many=True).data
```
